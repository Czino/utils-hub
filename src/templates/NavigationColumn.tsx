import { useEffect, useState } from 'react'
import { FiMenu, FiX } from 'react-icons/fi'
import { useToggle } from '../hooks/useToggle'

export const navigation = [
  { path: '/utils/list', title: 'List' },
  { path: '/utils/text', title: 'Text' },
  { path: '/utils/units', title: 'Units' },
  { path: '/utils/fractions', title: 'Fractions' },
  { path: '/utils/date', title: 'Dates' },
] as const

export const NavigationColumn = () => {
  const [isVisible, toggleIsVisible] = useToggle()
  const [current, setCurrent] = useState('')

  useEffect(() => {
    setCurrent(window.location.pathname)
  }, [])

  return (
    <>
      <nav
        className={[
          'bg-white absolute p-4 transition-[top]',
          'w-full h-full',
          isVisible ? 'top-12' : 'top-full',
          'md:h-screen md:relative md:left-0 md:w-[240px]',
        ].join(' ')}
      >
        <ul>
          <li className="font-bold">
            <a href="/">UtilityHub</a>
          </li>
          {navigation.map(({ path, title }) => (
            <li key={path}>
              <a href={path} className={current.includes(path) ? 'text-highlight-1' : ''}>
                {title}
              </a>
            </li>
          ))}
        </ul>
      </nav>
      <div
        className="flex absolute right-0 bottom-0 justify-end p-4 w-full bg-white md:hidden"
        role="button"
        onClick={toggleIsVisible}
      >
        {isVisible ? <FiX aria-label="close navigation" /> : <FiMenu aria-label="open navigation" />}
      </div>
    </>
  )
}
