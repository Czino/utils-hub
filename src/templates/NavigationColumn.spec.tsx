import '@testing-library/jest-dom'
import { render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { toMatchDiffSnapshot } from 'snapshot-diff'
import { NavigationColumn, navigation } from './NavigationColumn'
expect.extend({ toMatchDiffSnapshot })

describe('NavigationColumn', () => {
  const { asFragment } = render(<NavigationColumn />)
  const base = asFragment()

  beforeEach(() => {
    Object.defineProperty(window, 'location', {
      configurable: true,
      enumerable: true,
      value: new URL(window.location.href),
    })
  })
  it('should render NavigationColumn', () => {
    expect(base).toMatchSnapshot()
  })
  it('should highlight current page', () => {
    const { path, title } = navigation[1]!
    window.location.href = `http://localhost:3000${path}`
    const { getByText } = render(<NavigationColumn />)

    expect(getByText(title)).toHaveClass('text-highlight-1')
  })
  it('should show NavigationColumn when clicking on menu icon', async () => {
    const { getByLabelText } = render(<NavigationColumn />)
    await userEvent.click(getByLabelText('open navigation'))

    expect(asFragment()).toMatchDiffSnapshot(base)
  })
})
