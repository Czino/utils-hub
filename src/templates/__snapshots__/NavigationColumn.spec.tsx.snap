// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationColumn should render NavigationColumn 1`] = `
<DocumentFragment>
  <div
    class="fixed inset-0 transition-[opacity] bg-black pointer-events-none opacity-0"
  />
  <nav
    class="bg-white fixed z-10 p-4 transition-[top] overflow-auto pb-32 w-full h-full top-full md:h-screen md:relative md:left-0 md:w-[240px] md:border-right-1"
  >
    <ul>
      <li
        class="font-bold"
      >
        <h1>
          <a
            href="/"
          >
            UtilityHub
          </a>
        </h1>
      </li>
      <li>
        <h2
          class="mt-4 font-bold"
        >
          Time
        </h2>
      </li>
      <li>
        <a
          class=""
          href="/utils/time/date"
        >
          Dates
        </a>
      </li>
      <li>
        <h2
          class="mt-4 font-bold"
        >
          Numbers
        </h2>
      </li>
      <li>
        <a
          class=""
          href="/utils/numbers/base"
        >
          Numbers 
        </a>
      </li>
      <li>
        <a
          class=""
          href="/utils/numbers/fractions"
        >
          Fractions
        </a>
      </li>
      <li>
        <a
          class=""
          href="/utils/numbers/percentages"
        >
          Percentages 
        </a>
      </li>
      <li>
        <a
          class=""
          href="/utils/units"
        >
          Units
        </a>
      </li>
      <li>
        <h2
          class="mt-4 font-bold"
        >
          Text
        </h2>
      </li>
      <li>
        <a
          class=""
          href="/utils/text"
        >
          Text
        </a>
      </li>
      <li>
        <a
          class=""
          href="/utils/text/list"
        >
          Lists
        </a>
      </li>
      <li>
        <a
          class=""
          href="/utils/text/encoding"
        >
          Encoding
        </a>
      </li>
      <li>
        <a
          class=""
          href="/utils/text/hash-functions"
        >
          Hash functions
        </a>
      </li>
    </ul>
  </nav>
  <div
    class="flex fixed right-0 bottom-0 z-20 justify-end p-4 w-full bg-white  md:hidden  border shadow-2xl border-top-1"
    role="button"
  >
    <svg
      aria-label="open navigation"
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <line
        x1="3"
        x2="21"
        y1="12"
        y2="12"
      />
      <line
        x1="3"
        x2="21"
        y1="6"
        y2="6"
      />
      <line
        x1="3"
        x2="21"
        y1="18"
        y2="18"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`NavigationColumn should show NavigationColumn when clicking on menu icon 1`] = `
"Snapshot Diff:
- First value
+ Second value

- <DocumentFragment />
+ <DocumentFragment>
+   <div
+     class="fixed inset-0 transition-[opacity] bg-black pointer-events-none opacity-0"
+   />
+   <nav
+     class="bg-white fixed z-10 p-4 transition-[top] overflow-auto pb-32 w-full h-full top-full md:h-screen md:relative md:left-0 md:w-[240px] md:border-right-1"
+   >
+     <ul>
+       <li
+         class="font-bold"
+       >
+         <h1>
+           <a
+             href="/"
+           >
+             UtilityHub
+           </a>
+         </h1>
+       </li>
+       <li>
+         <h2
+           class="mt-4 font-bold"
+         >
+           Time
+         </h2>
+       </li>
+       <li>
+         <a
+           class=""
+           href="/utils/time/date"
+         >
+           Dates
+         </a>
+       </li>
+       <li>
+         <h2
+           class="mt-4 font-bold"
+         >
+           Numbers
+         </h2>
+       </li>
+       <li>
+         <a
+           class=""
+           href="/utils/numbers/base"
+         >
+           Numbers 
+         </a>
+       </li>
+       <li>
+         <a
+           class=""
+           href="/utils/numbers/fractions"
+         >
+           Fractions
+         </a>
+       </li>
+       <li>
+         <a
+           class=""
+           href="/utils/numbers/percentages"
+         >
+           Percentages 
+         </a>
+       </li>
+       <li>
+         <a
+           class=""
+           href="/utils/units"
+         >
+           Units
+         </a>
+       </li>
+       <li>
+         <h2
+           class="mt-4 font-bold"
+         >
+           Text
+         </h2>
+       </li>
+       <li>
+         <a
+           class=""
+           href="/utils/text"
+         >
+           Text
+         </a>
+       </li>
+       <li>
+         <a
+           class=""
+           href="/utils/text/list"
+         >
+           Lists
+         </a>
+       </li>
+       <li>
+         <a
+           class=""
+           href="/utils/text/encoding"
+         >
+           Encoding
+         </a>
+       </li>
+       <li>
+         <a
+           class=""
+           href="/utils/text/hash-functions"
+         >
+           Hash functions
+         </a>
+       </li>
+     </ul>
+   </nav>
+   <div
+     class="flex fixed right-0 bottom-0 z-20 justify-end p-4 w-full bg-white  md:hidden  border shadow-2xl border-top-1"
+     role="button"
+   >
+     <svg
+       aria-label="open navigation"
+       fill="none"
+       height="1em"
+       stroke="currentColor"
+       stroke-linecap="round"
+       stroke-linejoin="round"
+       stroke-width="2"
+       viewBox="0 0 24 24"
+       width="1em"
+       xmlns="http://www.w3.org/2000/svg"
+     >
+       <line
+         x1="3"
+         x2="21"
+         y1="12"
+         y2="12"
+       />
+       <line
+         x1="3"
+         x2="21"
+         y1="6"
+         y2="6"
+       />
+       <line
+         x1="3"
+         x2="21"
+         y1="18"
+         y2="18"
+       />
+     </svg>
+   </div>
+ </DocumentFragment>"
`;
